### PR TITLE
DAOS-8931 vea: optimize free extent insertion

### PR DESCRIPTION
**Pull request by NiuYawei**
Inserting a free extent into free extent tree requires at most 4
tree lookups in worst cases (when the inserting free extent can't
be merged with other existing ones):

1. Lookup 'prev' extent(having smaller offset) to see if the inserting
   one could be merged with it;
2. Lookup 'next' extent(having larger offset) to see if the iserting
   one could be merged with it;
3. Insert the extent when it failed to be merged;
4. Fetch the newly inserted one to get the in-tree record for later
   operations;

When space is very fragmented, the tree lookup overhead will be
non-negligible and that could result in inefficient block reclaiming.

To reduced tree lookups, following dbtree APIs are introduced/improved:

- dbtree_fetch_cur()
  Fetch from the current tree trace position;

- dbtree_fetch_prev()
  Fetch from the prev position, caller may specify if the trace should
  be moved on fetch or not;

- dbtree_fetch_next()
  Fetch from the next position, caller may specify if the trace should
  be moved on fetch or not;

- dbtree_upsert():
  This API is improved to return the in-tree record address.

By using the new dbtree APIs, only 1 tree lookup is required on free
extent inserting now (the 1st step):

1. Lookup the free extent to be inserted. (The trace will be set to the
   position for later potential inserting);
2. dbtree_fetch_prev() to see if it could be merged with 'prev';
3. dbtree_fetch_cur() or dbtree_fetch_next() to see if it could be merged
   with 'next';
4. Insert the extent in current position when it failed to be merged;
5. Use the in-tree record returned by dbtree_upsert() for later operations;

Following are the vea_stress test results (default option, same rand seed):

without patch
Operation   Samples      Time(us)     Min(us)    Max(us)    Avg(us)
reserv         10629991     18547433     0          7230       1
tx_publish  1797397        6067054      0          599        3
tx_free        244712        22683884     0          5362       92
tx_merge    1477242      50602993     0          2573       34

with path
Operation   Samples      Time(us)     Min(us)    Max(us)    Avg(us)
reserv        11752846      19144825     0          11776      1
tx_publish  1986759       6703297      0           623        3
tx_free         273184       21447843     0          4994       78
tx_merge    1637848      47905031     0          2164       29

We can see tx_free avg latency reduced 15%, and 10%~12% more operations
are performed within same amount of time.

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]

File: src/vea/vea_free.c [hunk 1]
```diff
@@ -159,21 +160,56 @@ merge_free_ext(struct vea_space_info *vsi, struct vea_free_extent *ext_in,
...
 
-	if (rc == -DER_NONEXIST)
-		goto done;	/* Merge done */
-	else if (rc)
-		return rc;	/* Error */
+	if (fetch_prev) {
+		rc = dbtree_fetch_prev(btr_hdl, &key_out, &val, false);
+		if (rc == -DER_NONEXIST) {
+			fetch_prev = false;
+			goto repeat;
```
Comment (Nasf-Fan) [inline src/vea/vea_free.c:188]: If we cannot find the prev one, then can we set some flag (before goto repeat) to skip dbtree_fetch_cur() when try to find next?

**Comment by NiuYawei** [comment]
Merge master to pick the fix of DAOS-8901

File: src/vea/vea_free.c (hunk 1 above)
Comment (Nasf-Fan) [reply src/vea/vea_free.c:188]: I see what happened under such case

**Review by Nasf-Fan — APPROVED** [review]

**Comment by NiuYawei** [comment]
Merge OID refactor changes

**Comment by NiuYawei** [comment]
merge master to pick up rank shuffling fix.

**Pull request merged**